File: app/src/main/java/com/denma/planeat/arch/viewmodels/SearchScreenViewModel.java
package com.denma.planeat.arch.viewmodels;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.ViewModel;

import com.denma.planeat.arch.repositories.MenuRepository;
import com.denma.planeat.arch.repositories.ResponseRepository;
import com.denma.planeat.models.local.FoodMenu;
import com.denma.planeat.models.remote.Recipe;
import com.denma.planeat.models.remote.Response;

import java.util.concurrent.Executor;

import javax.inject.Inject;

public class SearchScreenViewModel extends ViewModel {

    // --- REPOSITORIES ---
    private MenuRepository menuDataSource;
    private ResponseRepository responseDataSource;
    private Executor executor;

    // --- CONSTRUCTOR ---
    @Inject
    SearchScreenViewModel(MenuRepository menuDataSource, ResponseRepository responseDataSource, Executor executor){
        this.menuDataSource = menuDataSource;
        this.responseDataSource = responseDataSource;
        this.executor = executor;
    }

    // FOR MENU
    // --- UPDATE ---
    public void updateMenu(final FoodMenu foodMenu){ executor.execute(() -> SearchScreenViewModel.this.menuDataSource.updateMenu(foodMenu));
    }

    // FOR RESPONSE
    // --- GET ---
    public LiveData<Response> getResponse(){ return this.responseDataSource.getResponse(); }

    // --- REMOTE DATA UPDATE ---
    public void updateResponseFromAPI(String query, String diet, String health){
        this.responseDataSource.updateResponseFromAPI(query, diet, health);
    }

    // FOR CURRENT RECIPE
    // --- CREATE ---
    public void setCurrentRecipe(Recipe recipe){this.responseDataSource.setCurrentRecipe(recipe); }

}

File: app/src/main/java/com/denma/planeat/arch/viewmodels/SearchScreenViewModel.kt
package com.denma.planeat.arch.viewmodels

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.ViewModel

import com.denma.planeat.arch.repositories.MenuRepository
import com.denma.planeat.arch.repositories.ResponseRepository
import com.denma.planeat.models.local.FoodMenu
import com.denma.planeat.models.remote.Recipe
import com.denma.planeat.models.remote.Response

import java.util.concurrent.Executor

import javax.inject.Inject

class SearchScreenViewModel
@Inject
internal constructor(private val menuDataSource: MenuRepository, private val responseDataSource: ResponseRepository, private val executor: Executor) : ViewModel() {

    // FOR RESPONSE
    // --- GET ---
    fun getResponse(): LiveData<Response> = this.responseDataSource.response

    // FOR MENU
    // --- UPDATE ---
    fun updateMenu(foodMenu: FoodMenu) = executor.execute { this.menuDataSource.updateMenu(foodMenu) }

    // --- REMOTE DATA UPDATE ---
    fun updateResponseFromAPI(query: String, diet: String?, health: String?) = this.responseDataSource.updateResponseFromAPI(query, diet, health)

    // FOR CURRENT RECIPE
    // --- CREATE ---
    fun setCurrentRecipe(recipe: Recipe) = this.responseDataSource.setCurrentRecipe(recipe)
}
